src/Covenant/Test.hs

```diff
@@ -69,27 +126,435 @@ instance Arbitrary Concrete where
+genArbitraryRecursive =
+  RecursiveConcreteDataDecl <$> do
+    tyNm <- freshTyName
+    baseCtor <- coerce <$> genConcreteConstructor -- any concrete ctor - or any ctor that doesn't contain the parent type - will suffice as a base case
+    numRecCtors <- GT.liftGen $ chooseInt (1, 5)
+    recCtor <- GT.vectorOf numRecCtors $ genRecCtor tyNm
+    returnDecl $ Covenant.Internal.Type.DataDeclaration tyNm count0 (Vector.fromList (baseCtor : recCtor))
+  where
+    genRecCtor :: Covenant.Internal.Type.TyName -> DataGenM (Covenant.Internal.Type.Constructor AbstractTy)
+    genRecCtor tyNm = do
+      ctorNm <- freshConstructorName
+      let thisType = Covenant.Internal.Type.Datatype tyNm Vector.empty
+      numNonRecArgs <- GT.liftGen $ chooseInt (1, 5) -- need at least one to avoid "bad" types
+      args <- coerce $ GT.vectorOf numNonRecArgs nestedCtorArg
+      pure $ Covenant.Internal.Type.Constructor ctorNm (Vector.fromList (thisType : args))
+
+{- Single variable polymorphic datatypes. That is, things like:
+
+   data Foo a = Nope | Yup a
+
+   data Snowk a = Start | More (Snowk a) a
+-}
+newtype Polymorphic1 = Polymorphic1 (Covenant.Internal.Type.DataDeclaration AbstractTy)
+  deriving (Eq) via (Covenant.Internal.Type.DataDeclaration AbstractTy)
+  deriving stock (Show)
+
+{- Generator for single variable polymorphic datatypes, no polymorphic *functions* as arguments to the datatypes yet (that requires something different).
+
+   When run multiple times in the monadic context, will reuse single variable declarations that are "in scope" (i.e. have already been generated and are
+   known in the DataGenM state).
+
+   TODO: Rework this to generate declarations with an arbitrary number of tyvar arguments. Doing so would be fairly simple (but isn't needed ATM)
+-}
+genPolymorphic1Decl :: DataGenM Polymorphic1
+genPolymorphic1Decl =
+  Polymorphic1 <$> do
+    tyNm <- freshTyName
+    logArity tyNm count1
+    numCtors <- GT.liftGen $ chooseInt (1, 5)
+    polyCtors <- concat <$> GT.vectorOf numCtors (genPolyCtor tyNm)
+    let result = Covenant.Internal.Type.DataDeclaration tyNm count1 (Vector.fromList polyCtors)
+    returnDecl result
+  where
+    -- We return a single constructor UNLESS we're generating a recursive type, in which case we have to return 2 to ensure a base case
+    genPolyCtor :: Covenant.Internal.Type.TyName -> DataGenM [Covenant.Internal.Type.Constructor AbstractTy]
+    genPolyCtor thisTy = do
+      ctorNm <- freshConstructorName
+      numArgs <- GT.liftGen $ chooseInt (1, 5)
+      argsRaw <- GT.vectorOf numArgs polyArg
+      let recCase = Covenant.Internal.Type.Datatype thisTy (Vector.singleton (Covenant.Internal.Type.Abstraction (BoundAt Z ix0)))
+      if recCase `elem` argsRaw
+        then do
+          baseCtorNm <- freshConstructorName
+          let baseCtor = Covenant.Internal.Type.Constructor baseCtorNm mempty
+              recCtor = Covenant.Internal.Type.Constructor ctorNm (Vector.fromList argsRaw)
```

Comment:
Slight nitpick, but since you know the length of `argsRaw`, it's more efficient to use `fromListN` in `GHC.Exts` here. Not relevant here due to how small `numArgs` is, but generally better practice with vectors.

Reply:
Noted, will try to remember to use that in the future